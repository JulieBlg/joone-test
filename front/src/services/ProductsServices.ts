export const getSpecificNumberOfProducts = async (numberOfProducts: string) => {
    const response = await window.fetch(`http://localhost:3000/products?limit=${numberOfProducts}`, {
        method: 'get',
        headers: { 'Content-Type': 'application/json' },
    });

    const { products } = await response.json();

    return products;
}

export default { getSpecificNumberOfProducts };