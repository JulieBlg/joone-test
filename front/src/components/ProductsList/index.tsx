import React, { FC, useEffect, useState } from 'react';

import { getSpecificNumberOfProducts } from '../../services/ProductsServices';

interface Product {
    title: string;
}

const ProductsList: FC = () => {
    const [products, setProducts] = useState([]);
    const [numberOfProducts, setNumberOfProducts] = useState('10');

    useEffect(() => {
        const fetchData = async () => {
            const data = await getSpecificNumberOfProducts(numberOfProducts);
            setProducts(data);
        }

        fetchData();
    }, [numberOfProducts])

    const handleChange = (event:React.ChangeEvent<HTMLInputElement>) => {
        setNumberOfProducts(event.target.value);
    }

    if(products.length === 0) return <>Aucun produit disponible</>
    
    return (
        <>
            <p>Nombre de produits Ã  afficher :</p>
            <input type="text" value={numberOfProducts} onChange={handleChange} />
            <ul>
            {products.map((product: Product) => {
                return <li>{product.title}</li>;
            })}
            </ul>
        </>
    );
}

export default ProductsList;