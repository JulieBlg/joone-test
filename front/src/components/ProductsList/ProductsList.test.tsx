import React from 'react';
import { render, findByText } from '@testing-library/react';
import ProductsList from '.';
import * as services from '../../services/ProductsServices';

let mockSpyGetProducts: jest.SpyInstance; 

beforeAll(() => {
    mockSpyGetProducts = jest.spyOn(services, 'getSpecificNumberOfProducts').mockResolvedValue(
        [
            { title: 'Lot de 3 couches' },
            { title: 'Lot de 5 couches' },
            { title: 'Lot de 10 couches' },
            { title: 'Lot de 20 couches' },
        ]
    );
})

it('renders a product list correctly', async () => {
    const { container, findByText } = render(<ProductsList />);

    expect(mockSpyGetProducts).toHaveBeenCalledWith('10');

    await findByText('Lot de 3 couches');

    expect(container).toMatchSnapshot();
});
