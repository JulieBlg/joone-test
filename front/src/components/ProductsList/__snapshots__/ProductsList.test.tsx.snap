// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a product list correctly 1`] = `
<div>
  <p>
    Nombre de produits Ã  afficher :
  </p>
  <input
    type="text"
    value="10"
  />
  <ul>
    <li>
      Lot de 3 couches
    </li>
    <li>
      Lot de 5 couches
    </li>
    <li>
      Lot de 10 couches
    </li>
    <li>
      Lot de 20 couches
    </li>
  </ul>
</div>
`;
